src/Repl/Repl.tsx

```diff
@@ -0,0 +1,317 @@
+// only print one board, but that is a much less common use case that the first one.
+/*
+const extendBoardsString = (board, value, boards) => {
+    let boardStr = decodeBoard(board["value"]);
+    if (boardStr !== value) boards = boards + boardStr + value + "\n";
+}
+*/
+
+
+/*
+ * Repl React function component
+ */
+const Repl = (props) => {
+
+  // Default is an empty buffer
+  const [inputBuffer, setInputBuffer] = useState(Array<any>());
+
+  // input mode: True is input mode, False is regular
+  // default is 'False'
+  const [inputMode, setInputMode] = useState(false);
+
+  // whether to report the type of the current expr
+  // default is 'False'
+  let shouldReportExprType = false;
+
+  // last expression we evaluated in the Repl
+  // when in Input mode, this is the expression we send, and the input buffer
+  // is updated instead
+  const [lastReplExpression, setLastReplExpression] = useState("");
+
+  /*
+   * enterInputMode
+   * Sets input mode true
+   */
+  const enterInputMode = (expr) => {
+    setInputMode(true);
+    setLastReplExpression(expr)
+  }
+
+  /*
+   * exitInputMode
+   * Clears the input buffer, set input mode false, clears replExpression
+   * Used after input has been fully accepted, or 'clear' has been entered
+   */
+  const exitInputMode = () => {
+    setInputBuffer([]);
+    setInputMode(false);
+    setLastReplExpression("")
+  }
+
+  /*
+   * Called when clear is pressed
+   */
+  const clear = () => {
+    exitInputMode();
+    return " 🤖 BoGL Says: Ok, exiting Input Mode. ";
+  }
+
+  /*
+   * pushToInputBuffer
+   * Pushes an input into the inputBuffer
+   *
+   * nextInput  : The next input to push in
+   */
+  const pushToInputBuffer = (nextInput: string) => {
+    let cpy = inputBuffer
+    cpy.push({"input":nextInput})
+    setInputBuffer(cpy)
+  }
+
+
+  /*
+   * popFromInputBuffer
+   * Pops the last input off the buffer, used when an error has occurred
+   */
+  const popFromInputBuffer = () => {
+    let cpy = inputBuffer
+    cpy.pop()
+    setInputBuffer(cpy)
+  }
+
+
+  /*
+   * handlePrompt
+   * Handles prompt for BoGL input
+   *
+   * decoded: The decoded response we can use
+   */
+  const handlePrompt = (replExpression,decoded) => {
+    let res = ""
+    if(inputMode === false) {
+      res += "\n 🤖 BoGL Says: Enter input, or \"clear\" to stop. \n";
+      enterInputMode(replExpression);
+    }
+
+    // only show a decoded board if there is one
+    if(decoded["boards"] && decoded["boards"] !== "") {
+      res = decodeValue(decoded["boards"]) + res
+    }
+
+    return res
+  }
+
+
+  /*
+   * handleTypeError
+   * Handles reporting a BoGL type error
+   *
+   * decoded: The decoded response we can use
+   */
+  const handleTypeError = (decoded) => {
+    let res = decoded["contents"]["message"]
+    if(inputMode === true) {
+      popFromInputBuffer();
+      res += "\nYour last input was discarded.";
```

Comment:
This can pulled out into a function or moved up into the caller in some way. There are 2 other duplicate cases below.

Reply:
Please do!